### PR TITLE
chore: [IAI-24] Upgrade `react-navigation` to `v5`

### PR DESCRIPTION
**Pull request by fabriziofff**
## Short description
This pr upgrades `react-navigation` from `v4` to `v5`.
Please refer to https://docs.google.com/document/d/1Bme0eHFIe431eR0Qc4KTiDrmjjW4hTsj0U_aXGcKjDA/ in order to have an overview of the changes made.

## List of changes proposed in this pull request
- Completely removed the logic for managing external deeplinks (not supported at the moment)
- Added `IONavigationContainer` that contains the `NavigationContainer` and `AppStackNavigator` (navigator entrypoint)
- Changed `ts/components/ui/Markdown/handlers/internalLink.ts` in order to map a route to a `NavigationAction`
- `useNavigationContext` has been removed and replaced with the new `useNavigation`

### Testing
- Added a new `./jestSetupAfterEnv.js` step, in order to execute before each test `jest.useFakeTimers();` and avoid https://github.com/facebook/jest/issues/6434
- Added just one time `jest.mock("react-native-device-info", () => mockRNDeviceInfo);` in `jestSetup.js`
- Updated e2e dev-server configuration `scripts/api-config.json` in order to check a payment.
- Changed `renderScreenFakeNavRedux` in order to support a screen or component that uses the compat navigator
- Added `renderScreenWithNavigationStoreContext` for the screen in the new navigators

## How to test
⛵ Navigate!

### CODE REVIEW

File: ts/components/services/ContactPreferencesToggles/index.tsx [hunk 1]
```diff
@@ -70,7 +70,7 @@ const ContactPreferencesToggle: React.FC<Props> = (props: Props) => {
     [serviceId, loadServicePreference]
   );
 
-  const nav = useNavigationContext();
+  const nav = useNavigation();
```
Comment (CrisTofani) [inline ts/components/services/ContactPreferencesToggles/index.tsx:73]: By using this new hook from navigation it looks that isFocused information is not changing properly, so, as an example, if we activate a bonus starting from its special service detail screen, when we navigate back and come back correctly to the service detail screen the preferences won't reload


https://user-images.githubusercontent.com/3959405/159503527-4fd821fc-1432-43f1-a666-6b7cb15a0075.mp4

Comment (CrisTofani) [reply ts/components/services/ContactPreferencesToggles/index.tsx:73]: [This](https://github.com/pagopa/io-app/blob/a63fcfd90078cdecdc5918dce8087b7aa64bc15a/ts/components/services/ContactPreferencesToggles/index.tsx#L77) is the hook i was mentioning

File: ts/screens/wallet/payment/ConfirmPaymentMethodScreen.tsx [hunk 2]
```diff
@@ -89,8 +91,11 @@ export type ConfirmPaymentMethodScreenNavigationParams = Readonly<{
   psps: ReadonlyArray<Psp>;
 }>;
 
-type OwnProps =
-  NavigationStackScreenProps<ConfirmPaymentMethodScreenNavigationParams>;
+type OwnProps = {
+  navigation: CompatNavigationProp<
+    IOStackNavigationProp<WalletParamsList, "PAYMENT_CONFIRM_PAYMENT_METHOD">
```
Comment (dgopsq) [inline ts/screens/wallet/payment/ConfirmPaymentMethodScreen.tsx:96]: Since the `navigation` in the props and the `useNavigation` hook must be typed as the same, what do you think about moving `IOStackNavigationProp<WalletParamsList, "PAYMENT_CONFIRM_PAYMENT_METHOD">` to a separate type alias?

File: ts/store/middlewares/navigation.ts [hunk 3]
```diff
@@ -1,32 +1,23 @@
...
-    setInstabugUserAttribute("activeScreen", screenName);
+  if (previousScreen !== currentScreen) {
+    setInstabugUserAttribute("activeScreen", currentScreen);
 
     // track only those events that are not included in the blacklist
-    if (!noAnalyticsRoutes.has(screenName)) {
-      void mixpanelTrack("SCREEN_CHANGE_V2", {
-        SCREEN_NAME: screenName
+    if (!noAnalyticsRoutes.has(currentScreen)) {
+      await mixpanelTrack("SCREEN_CHANGE_V2", {
```
Comment (dgopsq) [inline ts/store/middlewares/navigation.ts:14]: Is there a reason about the `await`? Can't we just execute this function non-blocking? Same for the other one.

File: ts/utils/__tests__/internalLink.test.ts [hunk 4]
```diff
@@ -21,6 +28,15 @@ describe("getInternalRoute", () => {
         IO_INTERNAL_LINK_PREFIX + validRoute + "?param1=value1&param2=value2",
         some({
           routeName: validRoute,
+          navigationAction: {
+            payload: {
+              name: "MAIN",
```
Comment (dgopsq) [inline ts/utils/__tests__/internalLink.test.ts:33]: Can we use constants here and in the other parts of this file?

File: ts/utils/navigation.ts [hunk 5]
```diff
@@ -1,71 +1,15 @@
...
-  }
-};
-
 /**
  * @deprecated
  */
 export const isOnboardingCompleted = () => {
-  const route = NavigationService.getCurrentState();
+  const route = navigationRef?.current?.getRootState();
+  // console.log(route);
```
Comment (dgopsq) [inline ts/utils/navigation.ts:10]: Can this comment be deleted or it's here for a reason?

**Review by dgopsq — CHANGES_REQUESTED** [review]
I added some comment and questions, mostly about the use of constants instead of hard-coded strings (just where we don't have autocomplete from the types) and some other minor stuff. I tried the application a bit and _I don't see broken behaviours_ but I don't really have the full know-how to test edge-cases 😟

File: ts/features/euCovidCert/screens/__test__/EUCovidCertValidScreen.test.tsx [hunk 6]
```diff
@@ -84,22 +84,40 @@ describe("Test EUCovidCertificateValidScreen", () => {
     if (detailsButton) {
       fireEvent.press(detailsButton);
       const detailsPayload: NavigationAction = {
-        type: "Navigation/NAVIGATE",
-        routeName: "EUCOVIDCERT_MARKDOWN_DETAILS",
-        params: { markdownDetails: "markdownDetails" }
+        type: "NAVIGATE",
+        payload: {
+          name: "MESSAGES_NAVIGATOR",
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/euCovidCert/screens/__test__/EUCovidCertValidScreen.test.tsx]: You're right, fixed in e4268782704b344d2194956ed782c8b9cae61707.
I'm going to replace all the strings with the consts! 
Btw, in the case of usage when we have the annotation with the navigation props, I think could be acceptable using directly the string, because in that case we have the typecheck and also the help of the IDE that suggests us only the available options, what you think?

File: ts/features/wallet/cobadge/component/__tests__/CobadgeWalletPreview.test.tsx [hunk 7]
```diff
@@ -136,10 +135,16 @@ describe("CobadgeWalletPreview component", () => {
...
     const cardComponent = component.queryByTestId("cardPreview");
     const expectedPayload: NavigationAction = {
-      type: NavigationActions.NAVIGATE,
-      routeName: ROUTES.WALLET_COBADGE_DETAIL,
-      params: {
-        cobadge: aCobadgeCard
+      type: "NAVIGATE",
+      // routeName: ROUTES.WALLET_COBADGE_DETAIL,
+      payload: {
+        name: "WALLET_NAVIGATOR",
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/wallet/cobadge/component/__tests__/CobadgeWalletPreview.test.tsx]: Fixed in ddfca6d069a991c39914661e71acda2032d5c53c

File: ts/features/zendesk/components/__tests__/ZendeskSupportComponent.test.tsx [hunk 8]
```diff
@@ -74,81 +86,72 @@ describe("the ZendeskSupportComponent", () => {
...
       it("if the assistanceForPayment is true should navigate to the ZendeskAskPermissions screen", () => {
         const store = createStore(appReducer, globalState as any);
         const component = renderComponent(store, true);
         const zendeskButton = component.getByTestId("contactSupportButton");
         fireEvent(zendeskButton, "onPress");
-        expect(navigateToZendeskAskPermissionsSpy).toBeCalled();
-        expect(navigateToZendeskPanicModeSpy).not.toBeCalled();
-        expect(navigateToZendeskChooseCategorySpy).not.toBeCalled();
+        expect(mockedNavigation).toHaveBeenCalledTimes(1);
+        expect(mockedNavigation).toHaveBeenCalledWith("ZENDESK_MAIN", {
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/zendesk/components/__tests__/ZendeskSupportComponent.test.tsx:100]: Fixed in 6e7c64dc6b9e2f69fd8292e873cc952ccf026175

File: ts/features/zendesk/screens/__tests__/ZendeskChooseCategory.test.tsx [hunk 9]
```diff
@@ -161,13 +173,15 @@ describe("the ZendeskChooseCategory screen", () => {
       );
       fireEvent(categoryItem, "onPress");
       expect(MockZendesk.addTicketCustomField).toBeCalled();
-      expect(navigateToZendeskChooseSubCategorySpy).toBeCalled();
+      expect(mockedNavigation).toHaveBeenCalledTimes(1);
+      expect(mockedNavigation).toHaveBeenCalledWith(
+        "ZENDESK_CHOOSE_SUB_CATEGORY",
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/zendesk/screens/__tests__/ZendeskChooseCategory.test.tsx]: Fixed in f7055a0513836015ae82f018e140bcb2da831da1!

File: ts/features/zendesk/screens/__tests__/ZendeskChooseSubCategory.test.tsx [hunk 10]
```diff
@@ -107,13 +118,15 @@ describe("the ZendeskChooseSubCategory screen", () => {
       );
       fireEvent(subCategoryItem, "onPress");
       expect(MockZendesk.addTicketCustomField).toBeCalled();
-      expect(navigateToZendeskAskPermissionsSpy).toBeCalled();
+      expect(mockedNavigation).toHaveBeenCalledWith("ZENDESK_ASK_PERMISSIONS", {
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/features/zendesk/screens/__tests__/ZendeskChooseSubCategory.test.tsx]: Fixed in 44d2822f17e448c83279b7ea4dcc68fffb31b3e6

File: ts/navigation/NavigationService.ts [hunk 11]
```diff
@@ -35,56 +35,45 @@ const withLogging =
...
 
 const dispatchNavigationAction = (action: NavigationAction) => {
-  navigator?.dispatch(action);
+  navigationRef.current?.dispatch(action);
 };
 
 const getCurrentRouteName = (): string | undefined =>
-  currentRouteState ? utilsGetCurrentRouteName(currentRouteState) : undefined;
+  navigationRef.current?.getCurrentRoute()?.name;
+// currentRouteState ? utilsGetCurrentRouteName(currentRouteState) : undefined;
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/navigation/NavigationService.ts]: Was the previous implementation that i kept for reference during the porting, but now we don't need them anymore!
Removed in 9c85b5a46d627ecc8ace0c3a481cc7e0b14fd488

File: ts/utils/navigation.ts (hunk 5 above)
Comment (fabriziofff) [reply ts/utils/navigation.ts]: Removed in c936f627b1f00002ae744346de2d231c51320400

File: ts/utils/__tests__/internalLink.test.ts (hunk 4 above)
Comment (fabriziofff) [reply ts/utils/__tests__/internalLink.test.ts]: Removed in 4f0c96984120de6802df1a45e18d42786fd523dd

File: ts/screens/wallet/payment/ConfirmPaymentMethodScreen.tsx (hunk 2 above)
Comment (fabriziofff) [reply ts/screens/wallet/payment/ConfirmPaymentMethodScreen.tsx]: Better for sure! Changed in 98b011359cd56d86ee36bbc272bbd415c96d1672

File: ts/store/middlewares/navigation.ts (hunk 3 above)
Comment (fabriziofff) [reply ts/store/middlewares/navigation.ts]: I agree, I don't think there is any added value having the async/await for the `trackScreen` method. 
Removed in 8abcc2a21f0287d9e11db9852857f9f3db5f42b9!

File: ts/navigation/AppStackNavigator.tsx [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        NavigationService.setNavigationReady();
+        routeNameRef.current = navigationRef.current?.getCurrentRoute()?.name;
+      }}
+      onStateChange={async () => {
+        const previousRouteName = routeNameRef.current;
+        const currentRouteName = navigationRef.current?.getCurrentRoute()?.name;
+
+        if (currentRouteName !== undefined) {
+          dispatch(setDebugCurrentRouteName(currentRouteName));
+          await trackScreen(previousRouteName, currentRouteName);
```
[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/navigation/AppStackNavigator.tsx]: Like the previous comment, I agree, removed in https://github.com/pagopa/io-app/commit/8abcc2a21f0287d9e11db9852857f9f3db5f42b9!

File: ts/features/euCovidCert/screens/__test__/EUCovidCertValidScreen.test.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgopsq) [reply ts/features/euCovidCert/screens/__test__/EUCovidCertValidScreen.test.tsx]: Yes 10/10 agree, I tried to add a comment only where we had hardcoded strings without typecheck (mostly in the tests). If in some comment I made there were the typecheck feel free to ignore it! 🙂

**Comment by debiff** [comment]
Beyond the conflicts, LGTM

**Review by CrisTofani — COMMENTED** [review]
LGTM

**Review by dgopsq — APPROVED** [review]
Same here ✌️

**Review by Undermaken — CHANGES_REQUESTED** [review]
I'm deeply testing these changes so I share what I found in while

### 💅🏻  double screens transition
- do logout
- can see double landing screen appearing in the transition

https://user-images.githubusercontent.com/822471/162253985-70731a51-78a7-43d4-beb2-9b8f9d9c8909.mov

### ⛔  new user can't move on TOS screen
- simulate [first onboarding](https://github.com/pagopa/io-dev-api-server/blob/58000b99dedb60e08bd617b884f4854a1dce4dc6/src/config.ts#L63)
- tos screen is shown
- user accepts
- the same tos screen is shown (🐞 )
- can't go ahead the TOS screen despite user accepts

https://user-images.githubusercontent.com/822471/162254331-03f6c1e0-737c-42f9-ab70-2df27bd6501c.mov

**Review by Undermaken — CHANGES_REQUESTED** [review]
when `usePaginatedMessages` is enabled

https://user-images.githubusercontent.com/822471/162412962-c7880f66-0a68-4b2e-8b62-8c0eca45c8cb.mov

**Review by Undermaken — CHANGES_REQUESTED** [review]
It seems some navigation params can't be serialized so there may be a leak of information

here what happens in the payment flow when the user try to add a payment method in while (same happens when the user tries to add a method starting from the wallet section)

I'm quite sure we use an `Option` to know if the adding of the payment method happens while a payment or not. Maybe isn't `Option` object not serializable?

### before (right behavior)
https://user-images.githubusercontent.com/822471/162416796-612e4c52-bb13-4498-9782-8662a1c019a2.mov

### now
https://user-images.githubusercontent.com/822471/162416882-e08d83eb-bcdb-46d9-bf60-5a7bcde60482.mov

here the warning

**Comment by fabriziofff** [comment]
@Undermaken here a recap:
**double screens transition:** If you agree, I think we can ignore this for now, since it is a minor glitch and only < 0,5% of users do the logout https://eu.mixpanel.com/s/YUx5y
**usePaginatedMessages**: Solved in 1219b414c5690e9b15f2bb5a3bc9f5b91946cb21
**Non-serializable values**: This should not have effect at the moment, since we don't use the deeplink and the state persistence https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state I've tracked the issue in https://pagopa.atlassian.net/browse/IAI-181 but I think can be solved in another iteration.

**Review by Undermaken — APPROVED** [review]

**Review by Undermaken — APPROVED** [review]
